<div style="position: relative;">
<img src="https://user-images.githubusercontent.com/7065401/98728503-5ab82f80-2378-11eb-9c79-adeb308fc647.png"></img>

<h1 style="color: white; position: absolute; top:27%; left:10%;">
     INE Bootcamp
</h1>
<h2 style="color: white; position: absolute; top:36%; left:10%;">
    Data Analysis, Visualization and Predictive Modeling
</h2> 

<h3 style="color: #ef7d22; font-weight: normal; position: absolute; top:58%; left:10%;">
    <b>David Mertz, Ph.D.</b>
</h3>

<h3 style="color: #ef7d22; font-weight: normal; position: absolute; top:63%; left:10%;">
    <b>Data Scientist</b>
</h3>
</div>

<div style="width: 100%; height: 200px; background-color: #222; text-align: center; padding-top: 20px; margin-bottom: 40px;">
<br><br>

<h1 style="color: white; font-weight: bold;">
    Basic curve fitting as predictive regression (solutions)
</h1>

<br><br> 
</div>

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
import seaborn as sns
sns.set_theme()

Let us look at some data for housing in King County, Washington (USA).

In [2]:
df = pd.read_csv('data/kc_house_data.csv')
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

For an illustration of polynomials, let us determine what single feature corresponds most strongly with price.

Including feature combinations is equivalent to fiting a high-dimensional polynomial.  This can often improve prediction further.

In [3]:
X = df.drop(columns=['id', 'zipcode', 'date',         # Clearly non-numeric
                     'lat', 'long', 'yr_renovated',   # Lat/lon "random"; yr_renovated often zero
                     'sqft_living15', 'sqft_lot15',   # Not clear distinction from base features
                     'price'])                        # Price HAS TO be excluded as target
y = df.price

<div style="width: 100%; height: 200px; background-color: #ef7d22; text-align: center; padding-top: 20px; margin-bottom: 40px;">
<br><br>

<h1 style="color: white; font-weight: bold;">
    Exercises
</h1>

<br><br> 
</div>

<h2 style="font-weight: bold;">
    Best feature evaluation
</h2>

![orange-divider](https://user-images.githubusercontent.com/7065401/98619088-44ab6000-22e1-11eb-8f6d-5532e68ab274.png)

Determine the model score for polynomial fits using only the top feature (including the linear fit)

In [4]:
# your solution here

We can find which features have the highest coefficents after training the model.  These are the ones with most influence on predictions.

In [5]:
X_poly = PolynomialFeatures(3).fit(X)
X_poly_new = X_poly.transform(X)
model = LinearRegression().fit(X_poly_new, y)
model.score(X_poly_new, y)

0.7464454507331919

In [6]:
poly_df = pd.DataFrame({'coefficient': model.coef_, 'column_pos':range(len(model.coef_))},
                       index=X_poly.get_feature_names(df.columns))
poly_df['magnitude'] = poly_df.coefficient.abs()
poly_df.sort_values('magnitude', inplace=True, ascending=False)
poly_df.drop('magnitude', axis=1, inplace=True)
poly_df

coefficient  column_pos
price bathrooms        -7.599539e+02          38
bathrooms grade        -6.960170e+02          62
bathrooms view         -6.489963e+02          60
price waterfront        2.785242e+02          42
date price             -2.415801e+02          26
...                              ...         ...
bedrooms^2 floors      -6.985817e-07         294
bedrooms sqft_lot view -6.407301e-07         317
id bedrooms^2           4.162498e-07         124
bedrooms^2 grade       -5.753365e-08         298
bedrooms^3             -3.791190e-12         290

[455 rows x 2 columns]

We can take only our favorite N columns of the total 455.  Let's say 100 features, for example.

In [7]:
column_pos = poly_df.column_pos.head(100)
column_pos.head()

price bathrooms     38
bathrooms grade     62
bathrooms view      60
price waterfront    42
date price          26
Name: column_pos, dtype: int64

In [8]:
X_poly_new[:, column_pos].shape

(21613, 100)

In [9]:
X_poly_select = X_poly_new[:, column_pos]
model = LinearRegression().fit(X_poly_select, y)
model.score(X_poly_select, y)

0.7224442307198866

<h2 style="font-weight: bold;">
    Best few features
</h2>

![orange-divider](https://user-images.githubusercontent.com/7065401/98619088-44ab6000-22e1-11eb-8f6d-5532e68ab274.png)

Determine the model score for polynomials using the top two or three features.

In [10]:
# your solution here

In [11]:
column_pos = poly_df.column_pos.head(3)
X_poly_three = X_poly_new[:, column_pos]
model = LinearRegression().fit(X_poly_three, y)
model.score(X_poly_three, y)

0.5019024307050435

---
<div style="position: relative; text-align: right;">
<img src="https://user-images.githubusercontent.com/7065401/98614301-dcf01780-22d6-11eb-9c8f-65ebfceac6f6.png" style="width: 130px; display: inline-block;"></img>

<img src="https://user-images.githubusercontent.com/7065401/98864025-08deda80-2448-11eb-9600-22aa17884cdf.png" style="height: 100%; max-height: inherit; position: absolute; top: 20%; left: 0px;"></img>
<br>

<h2 style="font-weight: bold;">
    David Mertz, Ph.D.
</h2>

<h3 style="color: #ef7d22; margin-top: 0.8em">
    Data Scientist
</h3>
<hr>
<br><br>

<p style="font-size: 80%; text-align: right; margin: 10px 0px;">
    david.mertz@gmail.com
</p>
<p style="font-size: 80%; text-align: right; margin: 10px 0px;">
    linkedin.com/in/dmertz/
</p>

</div>

<br><br><br>